engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/TailCall.scala

```diff
@@ -533,4 +141,344 @@ case object TailCall extends IRPass {
         )
       )
   }
+
+  private class Mini(private val isInTailPos: Boolean) extends MiniIRPass {
+    private val tailCandidates = {
+      val map = new java.util.IdentityHashMap[IR, Boolean]();
+      if (isInTailPos) {
+        map.put(null, true)
+      }
+      map
+    }
+
+    override def transformModule(m: Module): Module = {
+      m.bindings.map(subTailModuleBinding)
+      m
+    }
+
+    override def prepare(parent: IR, child: Expression): Mini = {
+      parent match {
+        case m: Module => m.bindings.map(subTailModuleBinding)
+        case e: Expression =>
+          if (tailCandidates.remove(null)) {
+            subTailExpression(e, true)
+            e.updateMetadata(TAIL_META)
+          }
+        case _ =>
+      }
+      val tailStatus = isTailPosition(child)
+      subTailExpression(child, tailStatus)
+      if (!tailStatus && isTailAnnotated(child)) {
+        child.addDiagnostic(
+          Warning.WrongTco(child.identifiedLocation())
+        )
+      }
+      this
+    }
+
+    override def transformExpression(ir: Expression): Expression = {
+      ir
+    }
+
+    private def isTailPosition(ir: Expression): Boolean = {
+      if (tailCandidates.containsKey(ir)) {
+        tailCandidates.get(ir)
+      } else {
+        // if null is in the candidates, replace it with first Expression
+        val initialState = tailCandidates.remove(null)
+        if (initialState) {
```

Comment:
Again, why not use just `if (isInTailPos)` why this need to remove `null` key from a hashmap?

Reply:
Fixed in ef3450c

Since `MiniIRPass.prepare` works on edges as of ffd27dfe9b6fbbd59cccb35633cb802be097815c there is a way choose particular `isInTailPos` status for each child `Expression` individually. Thus there is no need for "pass global" `tailCandidates` map and we can use two `Mini.IN_TAIL_POS` and `Mini.NOT_IN_TAIL_POS`  singletons.